src/compiler/program.ts

```diff
@@ -2400,6 +2543,70 @@ namespace ts {
         }
     }
 
+    function parseConfigHostFromCompilerHost(host: CompilerHost): ParseConfigHost {
+        return {
+            fileExists: host.fileExists,
+            readDirectory: () => [],
+            readFile: host.readFile,
+            useCaseSensitiveFileNames: host.useCaseSensitiveFileNames()
+        };
+    }
+
+    export function getProjectReferenceFileNames(host: CompilerHost, rootOptions: CompilerOptions): string[] | undefined {
+        if (rootOptions.references === undefined) {
+            return [];
+        }
+
+        const result: string[] = [];
+        for (const ref of rootOptions.references) {
+            const refPath = resolveProjectReferencePath(host, rootOptions.configFilePath, ref);
+            if (!refPath || !host.fileExists(refPath)) {
+                return undefined;
+            }
+            result.push(refPath);
+        }
+        return result;
+    }
+
+    function resolveProjectReferencePath(host: CompilerHost, referencingConfigFilePath: string, ref: ProjectReference): string | undefined {
+        const rootPath = referencingConfigFilePath ? getDirectoryPath(normalizeSlashes(referencingConfigFilePath)) : host.getCurrentDirectory();
+        let refPath = combinePaths(rootPath, ref.path);
+        if (!host.fileExists(refPath)) {
+            refPath = combinePaths(refPath, "tsconfig.json");
+        }
+        return refPath;
+    }
+
+    export function walkProjectReferenceGraph(host: CompilerHost, rootOptions: CompilerOptions,
+        callback: (resolvedFile: string, referencedProject: CompilerOptions, settings: ProjectReference) => void,
+        errorCallback?: (failedLocation: string) => void) {
+        if (rootOptions.references === undefined) return;
+
+        const references = getProjectReferenceFileNames(host, rootOptions);

```

Comment:
we are going to call resolveProjectReferencePath on every entry a few lines after this, why do we need this pre pass?